Data preprocessing

zupon opened the pull request:
New file `preprocess.py` handles data preprocessing, using different functions for the different data sources. 

Modifies `normalizer.py` to add flags for language, data source, and what kind of normalization to do (token-based or sentence-based).

Adds `normalize_all.sh` script to loop over all our languages and data sources, running the normalizer over all of them.

zupon committed:
added af data and preprocess.py to new branch

zupon committed:
filled in methods for preprocessing different sources of data

zupon committed:
started adding lang/data info as flags

zupon committed:
added bash script to normalize all combos of lang and data

zupon committed:
moved language and data_source flags to else condition

zupon committed:
fix how UD splits line, fix i+=1 in binary for index

zupon committed:
merge master

zupon committed:
removed af data from remote repo

zupon committed:
work on normalizer.py

zupon committed:
Merge branch 'master' into data-preprocessing

zupon committed:
ran pylint over preprocess.py and tidied it up

sandyritchie commented on normalizer.py:
```diff
@@ -38,16 +42,37 @@ def main(argv):
+        try:
+            input_text: List[str] = preprocess.process_data(infile, FLAGS.data_source)
+        except Exception:
+            print("No data file from '{}' for '{}'".format(data_source, FLAGS.language))
```
Consider using f-strings.

zupon replied:
Good idea! I hadn't heard of f-strings before, but they are definitely cleaner than print .format.

sandyritchie reviewed (commented):
Can you add some tests for this new behavior?

crewee commented on preprocess.py:
```diff
@@ -0,0 +1,153 @@
+
+
+def substitute_brackets(string: str) -> str:
+    """Substitutes square brackets to work in Pynini."""
```
Could you add args and returns to these docstrings?

zupon replied:
Done!

crewee replied:
Thanks!

crewee commented on normalize_all.sh:
```diff
@@ -0,0 +1,12 @@
+for language in af mg
+do
+
+    for data_source in ud um lcc
```
Does this not include all of the data sources?

zupon replied:
Good catch. It looks like I only added a few data sources (and only two langauges) to make sure it worked and then forgot to go back and add them all.

crewee replied:
Oh haha, nice!  Thanks!

crewee commented on normalize_all.sh:
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+# For loop to normalize files from each data source for each language.
+
+for language in af mg
```
Does this also not include all the languages?

zupon replied:
Definitely an oversight! This was my minimal working example and then I forgot to go back in and add the rest.

crewee replied:
Nice!

crewee reviewed (changes requested):
Looking good!!

zupon committed:
changed print .format to f-strings

zupon committed:
added all languages and data sources to normalize_all.sh

zupon committed:
added docstring args to helper functions

zupon committed:
added tests for preprocessing files

zupon committed:
fixed split lines for ac data

zupon committed:
fixed tests that broke due to capitalization

zupon commented:
> Can you add some tests for this new behavior?

@sandyritchie I added a couple of tests for this:
- One that tries to preprocess a file that exists
- One that tries to preprocess a file that does not exist

What other tests would be useful here?
(also @crewee)

crewee commented:
> > Can you add some tests for this new behavior?
> 
> @sandyritchie I added a couple of tests for this:
> 
> * One that tries to preprocess a file that exists
> * One that tries to preprocess a file that does not exist
> 
> What other tests would be useful here?
> (also @crewee)

It might be cool to have an end to end test.  You could have a testfile that's essentially like a mock data file, then another test file that's an "expected", how we expect it to be normalized, then a test that shows that when we preprocess and normalize the mock data file, it ends up how we "expect" it to?  WDYT?

zupon committed:
added end to end test on dummy test file

sandyritchie commented:
> > > Can you add some tests for this new behavior?
> > 
> > 
> > @sandyritchie I added a couple of tests for this:
> > 
> > * One that tries to preprocess a file that exists
> > * One that tries to preprocess a file that does not exist
> > 
> > What other tests would be useful here?
> > (also @crewee)
> 
> It might be cool to have an end to end test. You could have a testfile that's essentially like a mock data file, then another test file that's an "expected", how we expect it to be normalized, then a test that shows that when we preprocess and normalize the mock data file, it ends up how we "expect" it to? WDYT?

Yeah dummy files would work, or even just an example input line from each data source and its expected output. I think I would also move the preprocess tests to a preprocess_test.py (also if normalizer_test.py is testing the normalizer_lib, I would call it normalizer_lib_test).

sandyritchie commented on preprocess.py:
```diff
@@ -0,0 +1,166 @@
+def process_ud_data(ud_file: str) -> List[str]:
+    """Processes UD conllu file into list of strings for normalization.
+
+    UD data format follows the CONLL data format, where each sentence includes
+    metadata and each token is on a separate line. Each line includes info
+    about the token, lemma, part of speech, and dependency syntax. For this
+    function we only care about the token.
```
An example line would help clarify this (and for the other data sources).

zupon replied:
I added some examples for each of the data sources. Some of them might need to be split across different lines though, since they exceed the maximum line length.

crewee commented:
> > > > Can you add some tests for this new behavior?
> > > 
> > > 
> > > @sandyritchie I added a couple of tests for this:
> > > 
> > > * One that tries to preprocess a file that exists
> > > * One that tries to preprocess a file that does not exist
> > > 
> > > What other tests would be useful here?
> > > (also @crewee)
> > 
> > 
> > It might be cool to have an end to end test. You could have a testfile that's essentially like a mock data file, then another test file that's an "expected", how we expect it to be normalized, then a test that shows that when we preprocess and normalize the mock data file, it ends up how we "expect" it to? WDYT?
> 
> Yeah dummy files would work, or even just an example input line from each data source and its expected output. I think I would also move the preprocess tests to a preprocess_test.py (also if normalizer_test.py is testing the normalizer_lib, I would call it normalizer_lib_test).

+1 great idea Sandy!

zupon committed:
added separate preprocessing tests

zupon committed:
changed normalizer_test to normalizer_lib_test

zupon committed:
added testdata directory with test files

zupon commented:
> > > > > Can you add some tests for this new behavior?
> > > > 
> > > > 
> > > > @sandyritchie I added a couple of tests for this:
> > > > 
> > > > * One that tries to preprocess a file that exists
> > > > * One that tries to preprocess a file that does not exist
> > > > 
> > > > What other tests would be useful here?
> > > > (also @crewee)
> > > 
> > > 
> > > It might be cool to have an end to end test. You could have a testfile that's essentially like a mock data file, then another test file that's an "expected", how we expect it to be normalized, then a test that shows that when we preprocess and normalize the mock data file, it ends up how we "expect" it to? WDYT?
> > 
> > 
> > Yeah dummy files would work, or even just an example input line from each data source and its expected output. I think I would also move the preprocess tests to a preprocess_test.py (also if normalizer_test.py is testing the normalizer_lib, I would call it normalizer_lib_test).
> 
> +1 great idea Sandy!

This all makes sense to me. I added some more tests to make sure we're preprocessing all the different data sources correctly (hint: we weren't). I fixed the preprocessing where things went wrong. I also did add some test files, since the formatting of some of the data sources makes it hard to add the one-line example in the tests, but the expected preprocessed output is just a line in the tests at this point.

I also did the renaming that @sandyritchie suggested for `normalizer_test` -> `normalizer_lib_test` and made the preprocessing tests separate.

sandyritchie commented on testdata/test_mg_ac_normalized.txt:
```diff
@@ -0,0 +1 @@
+amin'ny
```
This is great! Even though it's a bit redundant, I think your preprocess test will be easier to read if you just put these input strings directly in the test file rather than reading them from these txt files (it's ok to keep them here as well for the eventual end-to-end test). Obviously the UD one is a bit long so keeping that one in the txt file SGTM.

zupon replied:
That makes sense. Yeah, the normalized text should be short enough and in an easy enough format to just put it directly in the test file. I think the only input that will need to be in a separate file is actually the UD data.

sandyritchie commented on preprocess_test.py:
```diff
@@ -0,0 +1,90 @@
+    def test_preprocess_ud(self):
+        'Test loading in a ud data file.'
+        infile = "testdata/test_wo_ud_input.txt"
+        try:
+            input_text: List[str] = preprocess.process_data(infile, "ud")
+        except Exception:
+            input_text = None
```
I'm not sure if you need to test IO in every unit test, the last one should be fine.

zupon replied:
That is true, since we _know_ that the other inputs exist, so we can simplify the other tests.

zupon committed:
removed normalized files from repo (can add direct to test file)

zupon committed:
removed unneeded try/except, added expected direct to test

zupon committed:
removed normalized files from BUILD

crewee reviewed (approved):
LGTM, great work Andrew!
